<a href="https://colab.research.google.com/github/lokeshkumarn/Notebooks/blob/master/BankTermDeposit.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import os
import numpy as np
import pandas as pd

#Visualizations
import matplotlib as mpl
from matplotlib import pyplot as plt

#Modeling
import sklearn

Load the Data

In [0]:
data = pd.read_excel('Bank Data for case study assignment.xlsx')

In [72]:
data.head()

age          job marital status   ... housing loan? Personal loan   y
0   30   unemployed         married  ...            no            no  no
1   33     services         married  ...           yes           yes  no
2   35   management          single  ...           yes            no  no
3   30   management         married  ...           yes           yes  no
4   59  blue-collar         married  ...           yes            no  no

[5 rows x 8 columns]

Exploratory Data Analysis and Statistical Analysis

In [73]:
data.columns

Index(['age', 'job', 'marital status ', 'education', 'credit default?',
       'housing loan?', 'Personal loan', 'y'],
      dtype='object')

Column Marital Status is having extra space in the end, just rename it without space in the end

In [0]:
data.rename(columns={'marital status ':'marital status'},inplace=True)

In [75]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1021 entries, 0 to 1020
Data columns (total 8 columns):
age                1021 non-null int64
job                1019 non-null object
marital status     1020 non-null object
education          1020 non-null object
credit default?    1020 non-null object
housing loan?      1019 non-null object
Personal loan      1019 non-null object
y                  1021 non-null object
dtypes: int64(1), object(7)
memory usage: 63.9+ KB


Total 1021 records in that Marital Status, Education, Credit Default having 1 missing data and Job,Housing Loan, Personal Loan has two missing data

First fill the missing  value for the job

In [76]:
data[data['job'].isnull()]

age  job marital status  ... housing loan? Personal loan    y
79     40  NaN        married  ...           yes            no   no
1004   42  NaN         single  ...           yes           yes  yes

[2 rows x 8 columns]

Taking the first record from above and looking for data whose attribute values are similar to that of the missing value and decide how to replace the missing value

In [78]:
data[(data['age']== 40) & 
     (data['marital status'] =='married') & 
     (data['education'] =='secondary') & 
     (data['credit default?'] =='no') & 
     (data['housing loan?'] =='yes') & 
     (data['Personal loan'] =='no') & 
     (data['y'] =='no') & 
     (data['job'].isnull()==False)]

age          job marital status  ... housing loan? Personal loan   y
270    40     services        married  ...           yes            no  no
482    40   technician        married  ...           yes            no  no
597    40   management        married  ...           yes            no  no
671    40   technician        married  ...           yes            no  no
684    40     services        married  ...           yes            no  no
710    40   technician        married  ...           yes            no  no
960    40   technician        married  ...           yes            no  no
1012   40  blue-collar        married  ...           yes            no  no

[8 rows x 8 columns]

Taking the Mode and imputing the missing value for job 

In [79]:
data[(data['age']== 40) & 
     (data['marital status'] =='married') & 
     (data['education'] =='secondary')& 
     (data['credit default?'] =='no') & 
     (data['housing loan?'] =='yes') & 
     (data['Personal loan'] =='no') & 
     (data['job'].isnull()==False)]['job'].mode()

0    technician
dtype: object

In [81]:
data[(data['age']== 40) & 
     (data['marital status'] =='married') & 
     (data['education'] =='secondary') & 
     (data['credit default?'] =='no') &
     (data['housing loan?'] =='yes') & 
     (data['Personal loan'] =='no')& 
     (data['job'].isnull()==True)]

age  job marital status  ... housing loan? Personal loan   y
79   40  NaN        married  ...           yes            no  no

[1 rows x 8 columns]

imputing the new value for job using the mod value of job whose attributes had similar values as that of the missing values

In [0]:
data.loc[(data['age']== 40) & 
         (data['marital status'] =='married') & 
         (data['education'] =='secondary')    & 
         (data['credit default?'] =='no') & 
         (data['housing loan?'] =='yes') & 
         (data['Personal loan'] =='no')
    & (data['job'].isnull()==True),'job']=\
data[(data['age']== 40) & 
     (data['marital status'] =='married') & 
     (data['education'] =='secondary')    & 
     (data['credit default?'] =='no') & 
     (data['housing loan?'] =='yes') & 
     (data['Personal loan'] =='no')    & 
     (data['job'].isnull()==False)]['job'].mode()[0]

In [83]:
data[data['job'].isnull()]

age  job marital status  ... housing loan? Personal loan    y
1004   42  NaN         single  ...           yes           yes  yes

[1 rows x 8 columns]

Data imputed, from the above it is clear that only one is missing right now and do the similar kind of operation for above record also

In [85]:
data[(data['age']== 42) & 
     (data['marital status'] =='single') & 
     (data['education'] =='secondary') & 
     (data['credit default?'] =='no') & 
     (data['housing loan?'] =='yes') & 
     (data['Personal loan'] =='yes') & 
     (data['y'] =='yes') & 
     (data['job'].isnull()==False)]

Empty DataFrame
Columns: [age, job, marital status, education, credit default?, housing loan?, Personal loan, y]
Index: []

No data available for similar values.So look in broader scope

Remove the Marital Status and check

In [92]:
data[(data['age']== 42) & 
     (data['education'] =='secondary') & 
     (data['credit default?'] =='no') & 
     (data['housing loan?'] =='yes') & 
     (data['Personal loan'] =='yes') & 
     (data['y'] =='yes') & 
     (data['job'].isnull()==False)]

Empty DataFrame
Columns: [age, job, marital status, education, credit default?, housing loan?, Personal loan, y]
Index: []

Remove Education also and check

In [93]:
data[(data['age']== 42) & 
     (data['credit default?'] =='no') & 
     (data['housing loan?'] =='yes') & 
     (data['Personal loan'] =='yes') & 
     (data['y'] =='yes') & 
     (data['job'].isnull()==False)]

Empty DataFrame
Columns: [age, job, marital status, education, credit default?, housing loan?, Personal loan, y]
Index: []

Remove credit default

In [94]:
data[(data['age']== 42) & 
     (data['housing loan?'] =='yes') & 
     (data['Personal loan'] =='yes') & 
     (data['y'] =='yes') & 
     (data['job'].isnull()==False)]

Empty DataFrame
Columns: [age, job, marital status, education, credit default?, housing loan?, Personal loan, y]
Index: []

Remove housing loan

In [95]:
data[(data['age']== 42) &  
     (data['Personal loan'] =='yes') & 
     (data['y'] =='yes') & 
     (data['job'].isnull()==False)]

Empty DataFrame
Columns: [age, job, marital status, education, credit default?, housing loan?, Personal loan, y]
Index: []

Remove the Personal loan also from condition

In [96]:
data[(data['age']== 42) &  
     (data['y'] =='yes') & 
     (data['job'].isnull()==False)]

age         job marital status  ... housing loan? Personal loan    y
215   42  management        married  ...           yes            no  yes

[1 rows x 8 columns]

Finally we got that the missing value based on the result from above. So lets impute the job value to the missing value

In [99]:
data[(data['age']== 42) &  
     (data['y'] =='yes') & 
     (data['job'].isnull()==False)]['job'].mode()[0]

'management'

In [100]:
data[(data['age']== 42) & 
     (data['marital status'] =='single') & 
     (data['education'] =='secondary') & 
     (data['credit default?'] =='no') & 
     (data['housing loan?'] =='yes') & 
     (data['Personal loan'] =='yes') & 
     (data['y'] =='yes') & 
     (data['job'].isnull()==True)]

age  job marital status  ... housing loan? Personal loan    y
1004   42  NaN         single  ...           yes           yes  yes

[1 rows x 8 columns]

In [0]:
data.loc[(data['age']== 42) & 
     (data['marital status'] =='single') & 
     (data['education'] =='secondary') & 
     (data['credit default?'] =='no') & 
     (data['housing loan?'] =='yes') & 
     (data['Personal loan'] =='yes') & 
     (data['y'] =='yes') & 
     (data['job'].isnull()==True),'job']=\
data[(data['age']== 42) &  
     (data['y'] =='yes') & 
     (data['job'].isnull()==False)
    ]['job'].mode()[0]

Lets check the whether the values are imputed to the job

In [102]:
data[data['job'].isnull()]

Empty DataFrame
Columns: [age, job, marital status, education, credit default?, housing loan?, Personal loan, y]
Index: []

In [103]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1021 entries, 0 to 1020
Data columns (total 8 columns):
age                1021 non-null int64
job                1021 non-null object
marital status     1020 non-null object
education          1020 non-null object
credit default?    1020 non-null object
housing loan?      1019 non-null object
Personal loan      1019 non-null object
y                  1021 non-null object
dtypes: int64(1), object(7)
memory usage: 63.9+ KB


No empty values in job, so lets move on to the next attribute housing loan

In [104]:
data[data['housing loan?'].isnull()]

age        job marital status  ... housing loan? Personal loan   y
109    27  housemaid        married  ...           NaN           NaN  no
1016   33   services        married  ...           NaN            no  no

[2 rows x 8 columns]

as the first row has missing values in personal loan as well, so lets focus on the second row

In [107]:
data[(data['age']==33) &
    (data['job']=='services') &
    (data['marital status']=='married') &
    (data['education']=='secondary') &
    (data['credit default?']=='no') &
    (data['housing loan?'].isnull()==False) &
    (data['Personal loan']=='no') &
    (data['y']=='no')]

age       job marital status  ... housing loan? Personal loan   y
1007   33  services        married  ...           yes            no  no

[1 rows x 8 columns]

lets impute the value based on the above result

In [106]:
data[(data['age']==33) &
    (data['job']=='services') &
    (data['marital status']=='married') &
    (data['education']=='secondary') &
    (data['credit default?']=='no') &
    (data['housing loan?'].isnull()==False) &
    (data['Personal loan']=='no') &
    (data['y']=='no')]['job'].mode()[0]

'services'

In [0]:
data.loc[(data['age']==33) &
    (data['job']=='services') &
    (data['marital status']=='married') &
    (data['education']=='secondary') &
    (data['credit default?']=='no') &
    (data['housing loan?'].isnull()) &
    (data['Personal loan']=='no') &
    (data['y']=='no'),'housing loan?']=\
data[(data['age']==33) &
    (data['job']=='services') &
    (data['marital status']=='married') &
    (data['education']=='secondary') &
    (data['credit default?']=='no') &
    (data['housing loan?'].isnull()==False) &
    (data['Personal loan']=='no') &
    (data['y']=='no')]['job'].mode()[0]

In [109]:
data[data['housing loan?'].isnull()]

age        job marital status  ... housing loan? Personal loan   y
109   27  housemaid        married  ...           NaN           NaN  no

[1 rows x 8 columns]